drivers/serial/uart_ns16550_port.tmpl

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2010, 2012-2015 Wind River Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifdef CONFIG_UART_NS16550_PORT_@NUM@
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+static void irq_config_func_@NUM@(struct device *port);
+#endif
+
+static const struct uart_ns16550_device_config uart_ns16550_dev_cfg_@NUM@ = {
+	.sys_clk_freq = DT_UART_NS16550_PORT_@NUM@_CLK_FREQ,
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.irq_config_func = irq_config_func_@NUM@,
+#endif
+
+#ifdef DT_UART_NS16550_PORT_@NUM@_PCP
+	.pcp = DT_UART_NS16550_PORT_@NUM@_PCP,
+#endif
+
+#if DT_UART_NS16550_PORT_@NUM@_PCIE
+	.pcie = true,
+	.pcie_bdf = DT_UART_NS16550_PORT_@NUM@_BASE_ADDR,
+	.pcie_id = DT_UART_NS16550_PORT_@NUM@_SIZE,
+#endif
+};
+
+static struct uart_ns16550_dev_data_t uart_ns16550_dev_data_@NUM@ = {
+	.port = DT_UART_NS16550_PORT_@NUM@_BASE_ADDR,
+	.baud_rate = DT_UART_NS16550_PORT_@NUM@_BAUD_RATE,
+	.options = CONFIG_UART_NS16550_PORT_@NUM@_OPTIONS,
+
+#ifdef DT_UART_NS16550_PORT_@NUM@_DLF
+	.dlf = DT_UART_NS16550_PORT_@NUM@_DLF,
+#endif
+};
+
+DEVICE_AND_API_INIT(uart_ns16550_@NUM@, DT_UART_NS16550_PORT_@NUM@_NAME,
+		    &uart_ns16550_init,
+		    &uart_ns16550_dev_data_@NUM@, &uart_ns16550_dev_cfg_@NUM@,
+		    PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
+		    &uart_ns16550_driver_api);
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+static void irq_config_func_@NUM@(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+#ifdef UART_NS16550_PCIE_ENABLED
+	if (DEV_CFG(dev)->pcie) {
+		unsigned int irq = DT_UART_NS16550_PORT_@NUM@_IRQ;
+
+		if (irq == PCIE_IRQ_DETECT) {
+			irq = pcie_wired_irq(DT_UART_NS16550_PORT_@NUM@_BASE_ADDR);
+		}
```

Comment:
Sort of a metaphor skew going on here: this file is a template designed to automatically generate code correctly given compile-time configuration.  But the test for PCIE_IRQ_DETECT is being done at runtime, when AFAICT there's be no loss of generality to do that statically given that we're doing the same kind of thing already in this very spot.